Utils/azuretools-core/src/com/microsoft/azuretools/core/mvp/model/function/AzureFunctionMvpModel.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * All rights reserved.
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+ * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+ * the Software.
+ *
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+ * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+package com.microsoft.azuretools.core.mvp.model.function;
+
+import com.microsoft.azure.management.Azure;
+import com.microsoft.azure.management.appservice.AppServicePlan;
+import com.microsoft.azure.management.appservice.FunctionApp;
+import com.microsoft.azure.management.resources.Subscription;
+import com.microsoft.azuretools.authmanage.AuthMethodManager;
+import com.microsoft.azuretools.azurecommons.helpers.NotNull;
+import com.microsoft.azuretools.core.mvp.model.AzureMvpModel;
+import com.microsoft.azuretools.core.mvp.model.ResourceEx;
+import rx.Observable;
+import rx.schedulers.Schedulers;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public class AzureFunctionMvpModel {
+
+    public static final String CANNOT_GET_WEB_APP_WITH_ID = "Cannot get Function App with ID: ";
+    private final Map<String, List<ResourceEx<FunctionApp>>> subscriptionIdToFunctionApps;
+
+    private AzureFunctionMvpModel() {
+        subscriptionIdToFunctionApps = new ConcurrentHashMap<>();
+    }
+
+    public static AzureFunctionMvpModel getInstance() {
+        return SingletonHolder.INSTANCE;
+    }
+
+    public FunctionApp getFunctionById(String sid, String id) throws IOException {
+        Azure azure = AuthMethodManager.getInstance().getAzureClient(sid);
+        FunctionApp app = azure.appServices().functionApps().getById(id);
+        if (app == null) {
+            throw new IOException(CANNOT_GET_WEB_APP_WITH_ID + id); // TODO: specify the type of exception.
+        }
+        return app;
+    }
+
+    public void deleteFunction(String sid, String appId) throws IOException {
+        AuthMethodManager.getInstance().getAzureClient(sid).appServices().functionApps().deleteById(appId);
```

Comment:
dup code as AuthMethodManager.getInstance().getAzureClient(sid).appServices().functionApps()

Reply:
extract to `getFunctionAppsClient`